specs/VEP-Milestone25-vector-database-ingestion/README.md

```diff
@@ -0,0 +1,167 @@
+
+This leaves them with the question.
+How do we populate the data?
+Steps they need to complete:
+1. Read data from confluence/jira
+2. Chunk into paragraphs(or something similar)
+3. Embed into vector space
+4. save Vector and paragraph in vector database
+5. remove old information. For example if we are scraping jira every hour and we are writing details to the vector database we need to make sure we clean up all embeddings/chunks which were generated from old versions of the ticket.
+
+We will build a datajob in VDK which reads data from confluence or jira and writes it to a postgres instance with PGVector enabled. A embedding model will be running on a different machine which will be exposed through an API.
+We will make requests to the API to create embeddings for us.
+
+After this datajob is running we will create a template from this in which we think customers will be able to adopt to meet their use cases.
+
+#### Benefits to customers:
+They will be able to follow our template to quickly create similar jobs. This gives customers the ability to get up and runnning quickly.
+
+## Requirements and goals
+1. There should be a single pipelines which given jira/confluence credentials can scrape the source
+2. it should chunk up the information, embed it and then save it
+3. The systems should be easily configurable
+   1. Read from different sources
+   2. Different chunks sizes
+   3. Different embedders
+   4. Extra columns saved in database
+4. There should be an example on how to build your own ingestion pipeline
+5. Only scraping new data and removing old data must be supported
+
+
+### Non goals
+1. This only populates information into a database that could be used by as RAG system. We don't handle stuffing the actual prompts.
+
+## High-level design
+![sequence_diagram.png](sequence_diagram.png)
+
+## API design
+The api design should be heavily influenced by: https://python.langchain.com/docs/modules/data_connection/vectorstores/.
+
+The table should have a structure like
+
+| embedding     | text chunk                  | document id | metadata |
+|---------------|-----------------------------|------------- |------------- |
+| [1,2,3,4,5,6] | in this document blah...    | 15 | {"groups_with_access": ["IT", "HR"]} |
+| [1,2,3,3,5,6] | other content from same doc | 15 | {"groups_with_access": ["IT", "HR"]} |
+
+
+I think the python code could look something like this.
+In it we:
+1. delete any files that have been removed since the last scrape
+2. Then in a transaction delete all information for a page and in the same transaction write all the new information page
+3. The embedding api is abstracted into it own class allowing users to easily provide their own embedding api
+
+
+
+
+```python
+PostgresInstance.delete(ConfluenceReader(credentials_for_confluence).find_removed_document(last_timestamp))
+raw_page_or_tickets : Iterable[Union[PageChanges]] = ConfluenceReader(credentials_for_confluence).load(last_timestamp)
```

Comment:
From The example is a bit hard for me to understand where VDK fits. 

You really can use only LangChain for this: 

```
# https://python.langchain.com/docs/integrations/document_loaders/confluence
loader = ConfluenceLoader(
    url="https://yoursite.atlassian.com/wiki", username="me", api_key="12345"
)
documents = loader.load(space_key="SPACE", limit=50)
split_docs = langchain_text_splitter.split_documents(pages)
db = langchain.PGVector.from_documents(
    embedding= OpenAIEmbeddings(),
    documents=split_docs,
```

So is VDK value in that it can templetize this? E.g ConfluenceBasedRAGLoaderTemplate which would include the above and logic for cleaning and of course incremental ingestion/updates? 

Or is it only in the incremental update .  In above example the role VDK is really hard for me to grasp.

Reply:
I would mostly focus on the incrmental updates and use langchain where we can. 
But also open to disagreements on that point.